Sources/SwiftBuildSupport/SwiftBuildSystem.swift

```diff
@@ -174,6 +184,174 @@ private final class PlanningOperationDelegate: SWBPlanningOperationDelegate, Sen
+            }
+            return task
+        }
+
+        mutating func started(target: SwiftBuild.SwiftBuildMessage.TargetStartedInfo) throws {
+            if targetsByID[target.targetID] != nil {
+                throw Diagnostics.fatalError
+            }
+            targetsByID[target.targetID] = target
+        }
+
+        mutating func target(for task: SwiftBuild.SwiftBuildMessage.TaskStartedInfo) throws -> SwiftBuild.SwiftBuildMessage.TargetStartedInfo? {
+            guard let id = task.targetID else {
+                return nil
+            }
+            guard let target = targetsByID[id] else {
+                throw Diagnostics.fatalError
+            }
+            return target
+        }
+    }
+
+    private func emitInfoAsDiagnostic(info: SwiftBuildMessage.DiagnosticInfo) {
+        let fixItsDescription = if info.fixIts.hasContent {
+            ": " + info.fixIts.map { String(describing: $0) }.joined(separator: ", ")
+        } else {
+            ""
+        }
+        let message = if let locationDescription = info.location.userDescription {
+            "\(locationDescription) \(info.message)\(fixItsDescription)"
+        } else {
+            "\(info.message)\(fixItsDescription)"
+        }
+        let severity: Diagnostic.Severity = switch info.kind {
+        case .error: .error
+        case .warning: .warning
+        case .note: .info
+        case .remark: .debug
+        }
+        self.observabilityScope.emit(severity: severity, message: "\(message)\n")
+
+        for childDiagnostic in info.childDiagnostics {
+            emitInfoAsDiagnostic(info: childDiagnostic)
+        }
+    }
+
+    func emitEvent(
+        _ message: SwiftBuild.SwiftBuildMessage
+    ) throws {
+        guard !self.logLevel.isQuiet else { return }
+        switch message {
+        case .buildCompleted(let info):
+            progressAnimation.complete(success: info.result == .ok)
+            if info.result == .cancelled {
+                self.delegate?.buildSystemDidCancel(self.buildSystem)
+            } else {
+                self.delegate?.buildSystem(self.buildSystem, didFinishWithResult: info.result == .ok)
+            }
+        case .didUpdateProgress(let progressInfo):
+            var step = Int(progressInfo.percentComplete)
+            if step < 0 { step = 0 }
+            let message = if let targetName = progressInfo.targetName {
+                "\(targetName) \(progressInfo.message)"
+            } else {
+                "\(progressInfo.message)"
+            }
+            progressAnimation.update(step: step, total: 100, text: message)
+            self.delegate?.buildSystem(self.buildSystem, didUpdateTaskProgress: message)
+        case .diagnostic(let info):
+            if info.appendToOutputStream {
+                emitInfoAsDiagnostic(info: info)
+            }
+        case .output(let info):
+            let parsedOutput = String(decoding: info.data, as: UTF8.self)
```

Comment:
One consideration here: I don't think output messages are guaranteed to be split at character boundaries. So for example, we could emit two output messages, with the last byte of one message being the first byte of a 2-byte character and the second message having the second byte of that character. And then this String(decoding:) call might mangle the character.

I think it's also possible we receive something like `(output message) -> (progress update message) -> (output message)` which could split the output at arbitrary points.

Maybe when we receive an output message, we should append to a per-task buffer, then only decode it as a string and emit the output when the taskComplete message arrives?

Reply:
Good point, I've found this to be the case with some light local testing as well -- I'll give the per-task buffer a shot!